dbms/src/Flash/ResourceControl/LocalAdmissionController.h

```diff
@@ -0,0 +1,173 @@
+// Copyright 2023 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Common/Exception.h>
+#include <Common/Logger.h>
+#include <Flash/Executor/toRU.h>
+#include <Flash/Pipeline/Schedule/Tasks/Task.h>
+#include <Flash/ResourceControl/MockLocalAdmissionController.h>
+#include <Flash/ResourceControl/TokenBucket.h>
+#include <common/logger_useful.h>
+#include <kvproto/resource_manager.pb.h>
+#include <pingcap/kv/Cluster.h>
+
+#include <atomic>
+#include <memory>
+#include <mutex>
+
+namespace DB
+{
+class LocalAdmissionController;
+
+class ResourceGroup final : private boost::noncopyable
+{
+public:
+    explicit ResourceGroup(const resource_manager::ResourceGroup & group_pb_)
+        : name(group_pb_.name())
+        , user_priority(group_pb_.priority())
+        , user_ru_per_sec(group_pb_.r_u_settings().r_u().settings().fill_rate())
+        , group_pb(group_pb_)
+        , cpu_time_in_ns(0)
+        , log(Logger::get("rg:" + group_pb_.name()))
+    {
+        const auto & setting = group_pb.r_u_settings().r_u().settings();
+        initStaticTokenBucket(setting.burst_limit());
+    }
+
+#ifdef DBMS_PUBLIC_GTEST
+    ResourceGroup(const std::string & group_name_, uint32_t user_priority_, uint64_t user_ru_per_sec_, bool burstable_)
+        : name(group_name_)
+        , user_priority(user_priority_)
+        , user_ru_per_sec(user_ru_per_sec_)
+        , burstable(burstable_)
+        , log(Logger::get("rg:" + group_name_))
+    {
+        initStaticTokenBucket(user_ru_per_sec_);
+    }
+#endif
+
+    ~ResourceGroup() = default;
+
+    enum TokenBucketMode
+    {
+        normal_mode,
+        degrade_mode,
+        trickle_mode,
+    };
+
+#ifndef DBMS_PUBLIC_GTEST
+private:
+#endif
+    // Priority of resource group set by user.
+    // This is specified by tidb: parser/model/model.go
+    static constexpr int32_t LowPriorityValue = 1;
+    static constexpr int32_t MediumPriorityValue = 8;
+    static constexpr int32_t HighPriorityValue = 16;
+
+    // Minus 1 because uint64 max is used as special flag.
```

Comment:
`uint64 max` as special flag or `uint64 max`>>4 as a special flag?

Reply:
Looks `uint64_max >> 4` is more reasonable, because user_priority cannot be zero. Changed

Reply:
But using `uint64 max` is more convenient to implement the comparator of priority_queue. No need to add special if branch to check special flag in the comparator.